sw/host/tests/usbdev/usbdev_stream/stream_test.cc

```diff
@@ -0,0 +1,574 @@
+//
+//   -c   check any retrieved data against expectations
+//   -d   specify a particular USB device by bus number and device address
+//   -r   retrieve data from device
+//   -s   send data to device
+//   -t   use serial ports (ttyUSBx) in preference to libusb Bulk Transfer
+//        streams for usbdev_stream_test
+//   -v   verbose reporting
+//   -z   perform suspend-resume signaling throughout the test
+//
+// Build without libusb dependency:
+//   eg. g++ -Wall -Werror -o stream_test *.cc
+
+#include "stream_test.h"
+
+#include <cassert>
+#include <cctype>
+#include <cerrno>
+#include <cinttypes>
+#include <cstdbool>
+#include <cstddef>
+#include <cstdint>
+#include <cstdio>
+#include <cstdlib>
+#include <cstring>
+#include <ctime>
+#include <iostream>
+#include <sys/time.h>
+
+#include "usb_device.h"
+#if STREAMTEST_LIBUSB
+#include "usbdev_int.h"
+#endif
+#include "usbdev_serial.h"
+#include "usbdev_utils.h"
+
+// Test properties
+//
+// 16MiB takes about 40s presently with no appreciable CPU activity on the CW310
+// (ie. undefined transmitted data, and no checking of received data) but ca.
+// 152s with LFSR generation and checking across all of the 11 streams possible.
+//
+// Note: in normal use such as regression tests, the stream signatures will
+//       override the specified transfer amount.
+constexpr uint32_t kTransferBytes = (0x10U << 20);
+
+// Has any data yet been received from the device?
+bool received = false;
+
+// Time of first data reception.
+uint64_t start_time = 0U;
+
+// Configuration settings for the test.
+TestConfig cfg(false,  // Not verbose
+               true,   // Retrieve data from the device
+               true,   // Check the retrieved data
+               true);  // Send modified data to the device
+
+static USBDevice dev;
+
+// State information for each of the streams.
+static USBDevStream *streams[STREAMS_MAX];
+
+// Parse a command line option and return boolean value.
+static bool GetBool(const char *p);
+
+// Construct a modified port name for the next stream.
+static void PortNext(char *next, size_t n, const char *curr);
+
+// Parse a command line option and return boolean value.
+bool GetBool(const char *p) {
+  return (*p == '1') || (tolower(*p) == 'y') || (*p == '\r') || (*p == '\n') ||
+         (*p == '\0');
+}
+
+// Parse a command line option, retrieving a byte and indicating
+// success/failure.
+bool GetByte(const char **pp, uint8_t& byte) {
+  const char *p = *pp;
+  if (isdigit(*p)) {
+    uint32_t n = 0u;
+    do {
+      n = (n * 10) + *p++ - '0';
+    } while (n < 0x100u && isdigit(*p));
+    if (n < 0x100u) {
+      byte = (uint8_t)n;
+      *pp = p;
+      return true;
+    }
+  }
+  return false;
+}
+
+// Parse a command line option specifying the bus number and device address.
+bool GetDevice(const char *p, uint8_t &busNumber, uint8_t &devAddress) {
+  return GetByte(&p, busNumber) && (*p++ == ':') && GetByte(&p, devAddress);
```

Comment:
You don't check that `p` is the end of the string

Reply:
Okay.